### ISSUE TITLE
Warn when using "open" imports

### ISSUE THREAD

**Issue by garyb** [purescript/purescript#1509]
Include a copy-paste-able fix in the warning message that provides a version of the import with the used members selectively listed.

I had been thinking we could change the syntax for open imports and require `import SomeModule (..)`, but upon reflection that's kinda arbitrary breakage - the warning that states its own fix should be sufficient.

**Comment by hdgarrood** [comment]
Sorry if I'm a bit late to this discussion, but I'm not convinced that open imports should always be discouraged. I remember reading that the GHC team think open imports are fine, and often preferable, if you're importing a module in the same package.

**Comment by garyb** [comment]
Yeah, importing within the same package is a good case for allowing them actually. I guess we could allow open imports of a module name with the same prefix as the current module, but that's not quite the same thing - it would at least work for cases where we're re-exporting a bunch of inner modules though.

Thoughts @paf31?

**Comment by paf31** [comment]
I really like the idea of warning on unqualified whole module imports, but only if we can suggest the minimal imports.

I think qualified imports are enough of a fix for package-internal whole imports, no?

**Comment by garyb** [comment]
Yeah, I think so. I had in mind something like the warning would suggest a list below a certain number of imports, and qualification if something is heavily used. This might be a super annoying warning though :smile:

**Comment by hdgarrood** [comment]
The other thing is that modules like `Data.Maybe`, `Data.Tuple`, `Data.Either`, and of course `Prelude` are pretty much always imported open and I think it might be annoying to have either to import those qualified or maintain an explicit import list.

If not being able to work out where an identifier comes from is a problem I'd rather we address that through tooling like `psc-ide`, with a "display which module this identifier comes from" feature or something.

**Comment by hdgarrood** [comment]
Has opting in or out of individual warnings been discussed anywhere, incidentally?

**Comment by michaelficarra** [comment]
@hdgarrood There are other use cases that benefit from explicit imports, such as browsing code on GitHub.

**Comment by hdgarrood** [comment]
That's true. If we eventually copy Hackage and have highlighted source code on Pursuit with hyperlinked identifiers, do you think that would solve that issue?

Even so, there are a fairly large set of extremely commonly used types and functions that I don't think need explicit imports.

The situation I am worried about is that I have something like, say, `import Data.Maybe (Maybe(..))` in my module, along with a bunch of other imports, and then I decide I want to use `maybe` too, and so I have to be continuously sifting through my import list to sort it out. I also can't see an easy way of having tooling automatically work out what needs to happen in this case.

**Comment by michaelficarra** [comment]
@hdgarrood Not if you want to see what's on master or some experimental branch.

Tooling can add to your imports list automatically; see IntelliJ/Java integration.

**Comment by hdgarrood** [comment]
Yes, but are people looking at what's on master or some experimental branch on github that regularly? In particular, how often do people do that compared to how often people would have to fiddle with their import list? I'm guessing the latter happens much much more often. It certainly does in my workflow, at least.

Re tooling: I'm sure that kind of tooling is possible, I'm just not sure how long it will take to get there. Once we have tooling that handles this situation easily, then I'd be a lot happier about this change. (relevant: https://github.com/kRITZCREEK/psc-ide/issues/72)

**Comment by paf31** [comment]
I consider most of the warnings in `psc` to be more like HLint suggestions than anything else. It's a shame they have to show up in the console when things get recompiled, in that sense. It'd be better to have a nice UI to present them in, in an IDE.

For something like Pulp on the console, maybe it'd be good to have a way to just hide warnings.

**Comment by hdgarrood** [comment]
> I consider most of the warnings in psc to be more like HLint suggestions than anything else

I'm not sure I follow: what does this mean? That there are situations where you might want to ignore them?

**Comment by garyb** [comment]
Interesting, I'd say our warnings are a little stronger than hlint- more "we suggest that you do this so your code isn't straight up confusing or error prone" rather than "it's probably better to do it this way". The same as GHC warnings vs hlint.

**Comment by garyb** [comment]
I also have concerns about hiding warnings: I'd be unhappy to release libraries that failed any of these tests, regardless of whether the compiler raises them (yes I know I already have done :smile:).

**Comment by paf31** [comment]
I guess HLint isn't really what I meant to say. Point is, I think of warnings as a tool for development. I sometimes write code which I know will give a warning because I want to see the output - e.g. type wildcards. Or I might write my imports lazily by copy and paste or something and rely on warnings to tell me what I did wrong. So I think all warnings are something which should be acted upon, but quite a lot of warnings are like the output from user queries, and I think of minimal import warnings like that. What I would like is to be principled enough to explicitly list all my imports, but if I have the ability to run a query for the minimal imports, that's a good alternative. What I'm saying is that once this is merged, I hope the attitude towards elaborating imports will change because it will become so much easier.

**Comment by garyb** [comment]
How about this: we do have two forms of open module import. The one we have now, which provides the warnings, and then one that allows open imports if you _really_ want them:

``` purescript
import Prelude (..)
```

This can still be abused, but I think there is definitely a place where open imports will be valuable. Can you imaging how annoying [maintaining Data.Lens](https://github.com/purescript-contrib/purescript-profunctor-lenses/blob/e6b0fabbeafb3cd1681c6427ea87229a893d6cc1/src/Data/Lens.purs) would be without them, for instance?

**Comment by hdgarrood** [comment]
@paf31 ok, I see what you mean. But what about the workflow when you want to use some identifier from a module you've already imported with an explicit import list? Wouldn't that still be annoying?

**Comment by hdgarrood** [comment]
So I've started using qualified imports a lot more in the last few days, and it means I don't have to fiddle with import lists nearly as much, and the code is arguably more readable too :) so my position has softened somewhat.

If we want open imports to not be the default, why not go the whole hog (we're still <1.0, after all) and copy Elm's syntax? I think it's a lot more sensible, actually. We could also make a tool to automatically translate imports from the old syntax to the new one to ease the transition.

**Comment by hdgarrood** [comment]
That is:
* `import Prelude` would do what `import qualified Prelude as Prelude` does now
* `import Prelude exposing (..)` would do what `import Prelude` does now
* `import Prelude exposing (add)` would do what `import Prelude (add)` does now
* `import Prelude as P` would do what `import qualified Prelude as P` does now

**Comment by garyb** [comment]
I think `exposing` is unnecessary, also we don't use `qualified` anymore :wink:

See above for my suggestion regarding `import X (..)` too - I definitely like that approach, just not so sure that `import X` meaning `import X as X` is very useful. If we do make it so importing is required and unimported qualified references will not be resolved then it makes a bit more sense, but I still prefer the idea of having a "dev mode" open import that warns with an explicit list of imports.

**Comment by michaelficarra** [comment]
@hdgarrood See https://github.com/purescript/purescript/pull/1504, which made the `qualified` keyword optional, effectively deprecating it.

**Comment by hdgarrood** [comment]
Oh cool thanks. Have we considered removing the redundant syntax then?

**Comment by garyb** [comment]
There's [a warning now](https://github.com/purescript/purescript/commit/42702e84a8a388cb17684570152ddeb5659b18bf), it'll be removed in 0.9.

**Comment by hdgarrood** [comment]
Ha, ok, I should have looked at the code I guess. Sorry :)

Anyway you've convinced me: :+1: for this and for the "dev mode" open import.

**Comment by garyb** [comment]
I only added the warning recently :smile: - the non-`qualified` syntax is available in 0.7.6, but there's no warning.

**Comment by cdepillabout** [comment]
> I still prefer the idea of having a "dev mode" open import that warns with an explicit list of imports.

This sounds really nice!

I usually specifically list my imports, but like @hdgarrood, I hate fooling around with them.  I wish GHC had this feature too....

**Comment by garyb** [comment]
It's now implemented in the branch for #1685, give it a try if you like :)

**Comment by garyb** [comment]
To clarify, the behaviour there is:

```
import X -- warns about anything used from X
import X (..) -- no warning
```
